File: backend/go-chi-hive/internal/cells/user/service.go
package user

import (
	"context"
	"fmt"

	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hive/internal/shared/errors"
)

type service struct {
	repo           Repository
	passwordHasher PasswordHasher
}

func NewService(repo Repository, passwordHasher PasswordHasher) Service {
	return &service{
		repo:           repo,
		passwordHasher: passwordHasher,
	}
}

func (s *service) Register(ctx context.Context, cmd RegisterCommand) (*User, error) {
	existingUser, err := s.repo.FindByEmail(ctx, cmd.Email)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if existingUser != nil {
		return nil, errors.NewAlreadyExistsError("Email already exists")
	}

	existingUser, err = s.repo.FindByUsername(ctx, cmd.Username)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if existingUser != nil {
		return nil, errors.NewAlreadyExistsError("Username already exists")
	}

	hashedPassword, err := s.passwordHasher.Hash(cmd.Password)
	if err != nil {
		return nil, errors.NewInternalError(fmt.Sprintf("failed to hash password: %v", err))
	}

	user := &User{
		Username: cmd.Username,
		Email:    cmd.Email,
		Password: hashedPassword,
		Bio:      "",
		Image:    nil,
	}

	if err := s.repo.Create(ctx, user); err != nil {
		return nil, errors.NewInternalError(err.Error())
	}

	return user, nil
}

func (s *service) Login(ctx context.Context, cmd LoginCommand) (*User, error) {
	user, err := s.repo.FindByEmail(ctx, cmd.Email)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if user == nil {
		return nil, errors.NewNotFoundError("User not found")
	}

	if err := s.passwordHasher.Compare(user.Password, cmd.Password); err != nil {
		return nil, errors.NewInvalidCredentialsError("Invalid email or password")
	}

	return user, nil
}

func (s *service) GetUser(ctx context.Context, id int64) (*User, error) {
	user, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if user == nil {
		return nil, errors.NewResourceNotFound("User", "id", id)
	}
	return user, nil
}

func (s *service) GetUserByUsername(ctx context.Context, username string) (*User, error) {
	user, err := s.repo.FindByUsername(ctx, username)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if user == nil {
		return nil, errors.NewResourceNotFound("User", "username", username)
	}
	return user, nil
}

func (s *service) UpdateUser(ctx context.Context, cmd UpdateUserCommand) (*User, error) {
	user, err := s.repo.FindByID(ctx, cmd.ID)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if user == nil {
		return nil, errors.NewResourceNotFound("User", "id", cmd.ID)
	}

	if cmd.Email != nil && *cmd.Email != user.Email {
		existingUser, err := s.repo.FindByEmail(ctx, *cmd.Email)
		if err != nil {
			return nil, errors.NewInternalError(err.Error())
		}
		if existingUser != nil {
			return nil, errors.NewAlreadyExistsError("Email already exists")
		}
	}

	if cmd.Username != nil && *cmd.Username != user.Username {
		existingUser, err := s.repo.FindByUsername(ctx, *cmd.Username)
		if err != nil {
			return nil, errors.NewInternalError(err.Error())
		}
		if existingUser != nil {
			return nil, errors.NewAlreadyExistsError("Username already exists")
		}
	}

	var hashedPassword *string
	if cmd.Password != nil {
		hp, err := s.passwordHasher.Hash(*cmd.Password)
		if err != nil {
			return nil, errors.NewInternalError(fmt.Sprintf("failed to hash password: %v", err))
		}
		hashedPassword = &hp
	}

	user.Update(cmd.Username, cmd.Email, cmd.Bio, cmd.Image, hashedPassword)

	if err := s.repo.Update(ctx, user); err != nil {
		return nil, errors.NewInternalError(err.Error())
	}

	return user, nil
}

func (s *service) GetProfile(ctx context.Context, username string, observerID *int64) (*Profile, error) {
	profile, err := s.repo.GetProfileByUsername(ctx, username, observerID)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if profile == nil {
		return nil, errors.NewResourceNotFound("Profile", "username", username)
	}
	return profile, nil
}

func (s *service) FollowUser(ctx context.Context, followerID int64, username string) (*Profile, error) {
	return s.changeFollowStatus(ctx, followerID, username, s.repo.Follow)
}

func (s *service) UnfollowUser(ctx context.Context, followerID int64, username string) (*Profile, error) {
	return s.changeFollowStatus(ctx, followerID, username, s.repo.Unfollow)
}

func (s *service) changeFollowStatus(
	ctx context.Context,
	followerID int64,
	username string,
	action func(ctx context.Context, followerID, followedID int64) error,
) (*Profile, error) {
	followedUser, err := s.repo.FindByUsername(ctx, username)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if followedUser == nil {
		return nil, errors.NewResourceNotFound("User", "username", username)
	}

	if err := action(ctx, followerID, followedUser.ID); err != nil {
		return nil, errors.NewInternalError(err.Error())
	}

	return s.repo.GetProfileByUsername(ctx, username, &followerID)
}

File: backend/python-fastapi-vertical/app/features/user/service.py
from typing import Optional

from features.user.domain import Profile, User
from features.user.repository import find_by_email, find_by_username, create, find_by_id, update, \
    get_profile_by_username, follow, unfollow
from shared.errors import (
    new_already_exists_error,
    new_not_found_error,
    new_resource_not_found,
)
from shared.security.password import hash_password, compare_password


async def register(username: str, email: str, password: str) -> User:
    existing_email = await find_by_email(email)
    if existing_email:
        raise new_already_exists_error("Email already exists")

    existing_username = await find_by_username(username)
    if existing_username:
        raise new_already_exists_error("Username already exists")

    hashed_password = hash_password(password)
    user = User(
        id=0,
        username=username,
        email=email,
        password=hashed_password,
        bio="",
        image=None,
    )

    await create(user)
    return user


async def login(email: str, password: str) -> User:
    user = await find_by_email(email)
    if not user:
        raise new_not_found_error("User not found")

    await compare_password(user.password, password)
    return user


async def get_user(user_id: int) -> User:
    user = await find_by_id(user_id)
    if not user:
        raise new_resource_not_found("User", "id", user_id)
    return user


async def get_user_by_username(username: str) -> User:
    user = await find_by_username(username)
    if not user:
        raise new_resource_not_found("User", "username", username)
    return user


async def update_user(
    user_id: int,
    username: Optional[str] = None,
    email: Optional[str] = None,
    password: Optional[str] = None,
    bio: Optional[str] = None,
    image: Optional[str] = None,
) -> User:
    user = await get_user(user_id)

    if email and email != user.email:
        existing_email = await find_by_email(email)
        if existing_email:
            raise new_already_exists_error("Email already exists")

    if username and username != user.username:
        existing_username = await find_by_username(username)
        if existing_username:
            raise new_already_exists_error("Username already exists")

    hashed_password = None
    if password:
        hashed_password = hash_password(password)

    user.update(
        username=username,
        email=email,
        bio=bio,
        image=image,
        password=hashed_password,
    )

    await update(user)
    return user


async def get_profile(username: str, observer_id: Optional[int] = None) -> Profile:
    profile = await get_profile_by_username(username, observer_id)
    if not profile:
        raise new_resource_not_found("Profile", "username", username)
    return profile


async def follow_user(follower_id: int, username: str) -> Profile:
    user_to_follow = await find_by_username(username)
    if not user_to_follow:
        raise new_resource_not_found("User", "username", username)

    await follow(follower_id, user_to_follow.id)
    return await get_profile(username, follower_id)


async def unfollow_user(follower_id: int, username: str) -> Profile:
    user_to_unfollow = await find_by_username(username)
    if not user_to_unfollow:
        raise new_resource_not_found("User", "username", username)

    await unfollow(follower_id, user_to_unfollow.id)
    return await get_profile(username, follower_id)
